### ISSUE TITLE
Git compiled laggy

### ISSUE THREAD

**Issue by photocyte** [arescentral/antares#154]
Hi there, thanks for working on Antares!  A true blast from the past...  The 0.7.3 release works great, recently played through the whole original scenario on a Retina MBP running Mac OS X 10.10.3.  Compiled the most recent github revision on the same laptop to see if I could play around with it (upping the ship limit etc).  Antares compiled and runs, but is quite laggy. Even running at 640x480, from the intro crawler onwards it is jumping.

**Comment by sfiera** [comment]
Hi! Thanks for the report. Would you mind trying a few different binaries/source packages?

* http://downloads.arescentral.org/Antares/Antares-Source-0.7.3.zip is the source package for the 0.7.3 release. If you compile it, and the result is fine, it means I've introduced a bug since 0.7.3. (If not, there's something about your machine's configuration, but realistically, I don't think that's likely)
* http://downloads.arescentral.org/Antares/Antares-0.8.0rc2.zip is (was? :smile:) about to be released as 0.8.0. If it's slow for you too, then I'll postpone the release while we look at this.

Note: there is a problem with the text crawls right now where the text moves smoothly but the pictures jump two pixels at a time. I assume you're experiencing in-game slowness too.

**Comment by photocyte** [comment]
Happy to help! So the freshly compiled 0.7.3 works perfectly, however the 0.8.0rc2 binary has the same issue I was seeing with the most recent revision I compiled off github.  For the 0.8.0rc2 binary I tried at 640x480 on windowed mode and on fullscreen. 640x480 fullscreen on the retina display was somewhat comical, but does work.  Both modes had the same stuttering issue.

**Comment by sfiera** [comment]
OK. Do you know how to do a git bisect? It will take a while, but it's the simplest way to pinpoint when a bug was introduced. The process is like this:

    $ cd antares  # the git repository
    $ git bisect master 0.7.3 # master is bad, 0.7.3 is good
    ### git automatically picks a commit to test
    $ git submodule update && ./configure && make run # rebuild
    ### mark the commit as good or bad:
    $ git bisect good # if it's OK
    $ git bisect bad  # if you experience the stuttering
    ### git picks a new commit to test. Rebuild, repeat, until:
    ### git says something like:
    The first bad commit is 0123456789abcdef:
    "Ruin everything forever" by Chris Pickel <sfiera@sfzmail.com>

I'd guess it'll take about ten rebuilds to pinpoint the commit that introduced the problem. Each one will take a while, so you can run them while you're doing something else and check in every once in a while.

**Comment by photocyte** [comment]
Hmm... So it hasn't been working so well, can't seem to get the older commits to compile.  Any tips?  Following the the guide here:http://git-scm.com/docs/git-bisect

$ cd antares 
$ git bisect start
$ git bisect bad #marking the most recent commit bad
$ git bisect v0.7.1 #couldn't go to v0.7.3 - doesn't exist? https://github.com/arescentral/antares/releases
$ git bisect good #mark it good
Bisecting: 223 revisions left to test after this (roughly 8 steps)
[e6b4b11d4842302b1d76e8588440d44d8683cb01] More cleanup.
$ git submodule update
$ ./configure
gyp: ext/gmock-gyp/gmock.gyp not found (cwd: /Users/tim/Source/git/antares)
gyp: ext/libmodplug-gyp/libmodplug.gyp not found (cwd: /Users/tim/Source/git/antares)
gyp: ext/rezin/rezin.gyp not found (cwd: /Users/tim/Source/tim/antares)
gyp: ext/libzipxx/libzipxx.gyp not found (cwd: /Users/tim/Source/git/antares)
gyp: ext/libpng-gyp/libpng.gyp not found (cwd: /Users/tim/Source/git/antares)
gyp: ext/libsfz/libsfz.gyp not found (cwd: /Users/tim/Source/git/antares)

**Comment by sfiera** [comment]
Did you clone a new repository to do the bisect in? You'll need to run `git submodule init` in new repositories before `git submodule update` will work properly. If that's not it, can I see the repo's `.gitmodules` and `.git/config`? (there shouldn't be any private information in the latter, but please check before you post)

I forgot to push the `v0.7.3` tag. It's there now, and a pull should get it for you; sorry for the bad advice. `stable` will also get you `v0.7.3`; it always points to the latest release.

**Comment by photocyte** [comment]
Thanks : ) So heres my best shot, still some compiling issues:

$ git clone https://github.com/arescentral/antares.git
$ cd antares
$ git submodule init
$ ./configure
##configure works, relatively verbose
$ make
##App compiles successfully, app opens successfully to menu screen.  Has the stuttering
$ git bisect start
$ git bisect bad
$ git bisect good v0.7.3
Bisecting: a merge base must be tested
warning: unable to rmdir ext/glfw: Directory not empty
warning: unable to rmdir ext/libsndfile-gyp: Directory not empty
[dc652080ad817bd898838ba581fce709a9e22025] Release Antares 0.7.3.
$ git submodule update
fatal: reference is not a tree: a9bbd34339be8d192342b793f362ddf3e7b2684c
Submodule path 'ext/libmodplug-gyp': checked out 'cbb985f357911f68f708c31ab7733221e05857c3'
Submodule path 'ext/libpng-gyp': checked out 'b21534c8d1bc6ac51799e01e67fb93e6f4ee40b1'
Submodule path 'ext/libsfz': checked out '13ccbb51052246002a81c81657446dd2e1200293'
Submodule path 'ext/rezin': checked out 'd8e75d808539850f99108f4a41932a0630172837'
Cloning into 'test'...
remote: Counting objects: 50630, done.
remote: Compressing objects: 100% (5839/5839), done.
remote: Total 50630 (delta 13179), reused 7656 (delta 7656), pack-reused 37135
Receiving objects: 100% (50630/50630), 57.91 MiB | 8.35 MiB/s, done.
Resolving deltas: 100% (45109/45109), done.
Checking connectivity... done
Submodule path 'test': checked out '56ae0582a37a043fa3848df890d353094d31724b'
Unable to checkout 'a9bbd34339be8d192342b793f362ddf3e7b2684c' in submodule path 'data'
$ ./configure
##No text output
$ make
##Compiles normally, some C++ warnings but thats to be expected
##App opens to window config screen, crashes after hitting "Start"
##Message in Mac OS X console: Antares[60919]: couldn't find resource "fonts/tactical"
##Compiled app Antares.app/Contents/Resources/fonts/tactical.json exists.
##Zipped .gitmodules .git/config & compiled app - see the link below

https://www.dropbox.com/s/yd1hxockb3q8fff/Antares-troubleshooting.zip?dl=0

**Comment by sfiera** [comment]
Ah, yeah, sorry. The following will fix that:

    $ cd data
    $ git remote add sfiera https://github.com/sfiera/antares-data
    $ git fetch sfiera
    $ cd ..

You'll notice the fetch command downloading a couple dozen megabytes of data. They aren't used, so I hid them, but old versions of the antares repository don't work without them.

While I'm thinking of it, there's also a command "git bisect skip". There are a couple of points in the history where you might not be able to compile the game, so that will mark the revision as neither good nor bad and let you keep going.

**Comment by photocyte** [comment]
Thanks : ) Here's the git bisect result!
94899feb308141f0a6bac2c948b68f320f6bfe1f is the first bad commit
commit 94899feb308141f0a6bac2c948b68f320f6bfe1f
Author: Chris Pickel <sfiera@sfzmail.com>
Date:   Sun May 24 17:00:48 2015 +0900

    Switch to VBOs and vertex attributes.

:040000 040000 ad5e52757dc8de026d58d018bf0e651c2883d571 bad5d516d05d11dc78c1ca5955667b635feb8b7c M	src

**Comment by sfiera** [comment]
Hrm. That's a tricky one. Switching to vertex attributes is the essential part of moving from the legacy OpenGL pipeline to the modern one, which was necessary to get Antares working on Linux.

Can you post the information from "About this Mac" (Model, Processor, Memory, Startup Disk, Graphics)? I'm wondering if you're using a model with a dedicated GPU.

Here's my theory: on my machine, integrated graphics make it cheaper to shuttle data back and forth to the GPU, but on yours, that's taking a lot more time. So, even though you're using a "better" graphics card, Antares isn't using it properly (I knew that aspect of Antares was poorly optimized, but had figured it wasn't needed yet).

If so, can you try Antares under integrated (Intel) graphics? From searching, [gfxCardStatus](https://gfx.io/) will let you switch that.

If that's not it, can you grab me a profile? To do that, open Instruments.app and choose Time Profiler. Select "Choose Target…" from the "All Processes" toolbar, and find Antares.app. Record a trace—start a demo and watch it for a bit—then save the trace and send it to me (uploading here is best, but I'm not sure if you can; if not, sfiera@sfzmail.com).

**Comment by photocyte** [comment]
Hey sfiera, you're exactly right, I'm using a MBP with a dedicated GPU (NVIDIA GeForce GT 650M).  The GPU does get triggered when Antares starts up on dynamic switching mode, however if I set it to be restricted to the integrated card through gfxCardStatus before starting it up, the stuttering is gone.  See here for the traces:

https://www.dropbox.com/s/50fg59zxugoz9am/Antares_traces.zip?dl=0

There is a trace from the first commit that went bad, as well as the most recent git commit (both on integrated graphics and Nvidia).  At ~10 seconds "start" is hit on the Antares launcher window, it goes through the standard bitmap graphics intro, then around 25 seconds it hits the red scrolling text which has the stuttering.  From 25-~45 seconds is the red scrolling text, then I switch it to demo mode and it plays "While the Iron is Hot" from the original scenario.

Here's the specs on my machine:
  Model Identifier:	MacBookPro10,1
  Processor Name:	Intel Core i7
  Processor Speed:	2.6 GHz
  Number of Processors:	1
  Total Number of Cores:	4
  L2 Cache (per Core):	256 KB
  L3 Cache:	6 MB
  Memory:	16 GB

Intel HD Graphics 4000:

  Chipset Model:	Intel HD Graphics 4000
  Type:	GPU
  Bus:	Built-In
  VRAM (Dynamic, Max):	1024 MB
  Vendor:	Intel (0x8086)
  Device ID:	0x0166
  Revision ID:	0x0009
  gMux Version:	3.2.19 [3.2.8]
  Displays:
Color LCD:
  Display Type:	Retina LCD
  Resolution:	2880 x 1800 Retina
  Retina:	Yes
  Pixel Depth:	32-Bit Color (ARGB8888)
  Main Display:	Yes
  Mirror:	Off
  Online:	Yes
  Built-In:	Yes

NVIDIA GeForce GT 650M:

  Chipset Model:	NVIDIA GeForce GT 650M
  Type:	GPU
  Bus:	PCIe
  PCIe Lane Width:	x8
  VRAM (Total):	1024 MB
  Vendor:	NVIDIA (0x10de)
  Device ID:	0x0fd5
  Revision ID:	0x00a2
  ROM Revision:	3688
  gMux Version:	3.2.19 [3.2.8]

**Comment by sfiera** [comment]
OK. Since there's a workaround, I think I'm going to proceed with the 0.8.0 release.

I've been working on what I thought was going to be a fix, but I don't think the problem I was predicting is borne out by your traces. I'll have a binary for you to test in a day or two, but I'm less hopeful now that it will be the fix needed :/

**Comment by photocyte** [comment]
Sure thing, happy to test

**Comment by sfiera** [comment]
Hey. I see it’s been a couple of years since I updated this bug :/

Unfortunately, I don’t think I can fix it without my own machine with discrete graphics, and I don’t have access to one. Do you still have the same machine, and see the same issue? If not, I may just close this out, since I don’t foresee any progress.

**Comment by photocyte** [comment]
I'll give a shot at testing it tonight, to see if the issue pops up.  From my memory the just keeping it on integrated graphics was a decent workaround.

**Comment by sfiera** [comment]
I found some [new information](http://supermegaultragroovy.com//2016/12/10/auto-graphics-switching/) that might let me disable switching. Try this:
http://downloads.arescentral.org/Antares/antares-mac-dev.zip

I’m curious how it fares in various states. My hope is that it improves performance in gfxCardStatus’s “Dynamic Switching” mode; I suspect it will be unchanged under the others.

If it improves performance in the single monitor case, and you have any ability to try an external monitor, I would be interested in that, too. I think it will improve performance on the first display, not on the second display, and mirroring is anyone’s guess :)
